### PR TITLE
Distributed `vmap` functionality for vectorization across split dimension 

### PR DESCRIPTION
**Pull request by mrfh92**
## Due Diligence

- General:
    - [x]  **title** of the PR is suitable to appear in the [Release Notes](https://github.com/helmholtz-analytics/heat/releases/latest)
- Implementation:
    - [x] unit tests: all split configurations tested
    - [x] unit tests: multiple dtypes tested
    - [x] documentation updated where needed

## Description

Issue/s resolved: #1383 
The implemented vmap-functionality is similar to the one of PyTorch. It allows vectorized application of PyTorch-functions to one or multiple DNDarrays with vectorization taking place along the respective split axes. 

An artificial example, how vmap does work, is found in the tests. 

## Type of change
new feature

#### Does this change modify the behaviour of other functions? If so, which?
no

### CODE REVIEW

File: heat/core/vmap.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    randomness : {'error', 'different', 'same'}, optional
+        Determines how to handle randomness in the function to be vmapped. This argument is directly passed to the underlying PyTorch vmaps;
+        see the corresponding PyTorch documentation for more information and the note below.
+        If 'error' (default), an error is raised if the function to be mapped contains randomness.
+    chunk_size : int, optional
+        The size of the chunks to use for the process-local computation.
+        If None (default), apply a single PyTorch vmap over the process-local chunks of data. If not None, then compute the process-local PyTorch vmap `chunk_size`
+        many samples at a time. Note that `chunk_size=1` is equivalent to computing the process-local PyTorch vmap's with a for-loop.
+        If you run into memory issues computing the vmap, please try a non-None chunk_size.
+
```
Comment (JuanPedroGHM) [inline heat/core/vmap.py:35]: Docstring and function declaration should include the signature of the returned function (if possible). Makes using vmap easier

File: heat/core/vmap.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+from .communication import MPI_WORLD
+
+__all__ = ["vmap"]
+
+
+def vmap(func, out_dims=0, randomness="error", *, chunk_size=None):
+    """
+    This function is used to apply a function to a DNDarray in a vectorized way.
+    `heat.vmap` return a callable that can be applied to DNDarrays.
+    Vectorization will automatically take place along the split axis/axes; therefore, unlike in PyTorch, there is no argument `in_dims`.
```
Comment (JuanPedroGHM) [inline heat/core/vmap.py:17]: Connection between "batch axis" and "split axis". Do all parameters need to have the same split axis?

**Review by JuanPedroGHM — DISMISSED** [review]

File: heat/core/vmap.py (hunk 1 above)
Comment (mrfh92) [reply heat/core/vmap.py:45]: added signature with type hints

File: heat/core/vmap.py (hunk 2 above)
Comment (mrfh92) [reply heat/core/vmap.py]: added corresponding comment to docstring

**Review by ClaudiaComito — DISMISSED** [review]
Thanks a lot @mrfh92 , and @JuanPedroGHM for reviewing.

**Comment by mrfh92** [comment]
@ClaudiaComito problem should be fixed now. In fact, we require PyTorch >= 2.0.0 for vmap as vmap is only available in these versions.

**Review by ClaudiaComito — APPROVED** [review]

**Comment by mrfh92** [comment]
@ClaudiaComito codecov was a bit lower than expected by adding the uncovered line (~-3%), so I improved coverage

**Review by ClaudiaComito — APPROVED** [review]

**Comment by mrfh92** [comment]
@ClaudiaComito @JuanPedroGHM thx for reviewing :+1:

**Pull request merged**